node/tracing-logging-and-metrics-for-sdk-and-core.md

```diff
@@ -0,0 +1,96 @@
+- Logging in trace context can be done with otel span events or by extracting the span context into a log event, it's unclear to me if otel spans are a good fit here, see [this discussion](https://github.com/open-telemetry/opentelemetry-specification/issues/67) for info on the subject.
+
+| Level              | Component | Trace context | Event                                                 |
+| ------------------ | --------- | ------------- | ----------------------------------------------------- |
+| `INFO`             | Worker    | None          | Worker state change                                   |
+| `DEBUG`            | Worker    | Activation    | Activation received with list of jobs and WF context  |
+| `ERROR`            | Worker    | Activation    | WF activation failed                                  |
+| `DEBUG`            | Workflow  | Activation    | WF activation concluded with list of commands         |
+| `DEBUG`            | Worker    | Activity Task | Activity task received (start or cancel) with context |
+| `DEBUG` or `ERROR` | Worker    | Activity Task | Activity task resolved                                |
+| `ERROR`            | Worker    | Any           | Unexpected error occurred                             |
+| `ERROR`            | Workflow  | Activation    | Unexpected error occurred                             |
+| `DEBUG`            | Worker    | None          | Got activity heartbeat                                |
+
+> TODO: Fill in this table with required Core logs
+
+### Solution
+
+- Add context to logs emitted from Worker where applicable
+- Pass span context as from Lang to Core in the `poll_activity_task` and `poll_workflow_activation` APIs
+- Add methods to the Core API for exporting logs, metrics and traces
+  - Special care should be taken when defining the types here, logs should be structured to support custom formatting on the Lang side
+  - Hook up instrumentation into Lang opentelemetry tracer and Worker logger
+- Inject the Worker logger interface into the isolate for internal use - make sure activation context is baked into the WF logger
+- Worker logger should add a `component` attribute (e.g. `worker | core | workflow`) in the log context so we know each log's origin
+
+### Issues
+
+- Propagating traces, logs and metrics from Core to Lang implies buffering since they run in different threads and communicate using a queue
+  - Could result in event loss
+    - Is this tollerable?
+    - Should we also support logging directly to console from Core to mitigate this issue?
+  - Logs from different sources are received out of order
+    - This might not be that big of an issue in production where log aggregators can reorder based on timestamp
+    - During development, we can stall log output in order to merge and chronologically sort the different logs sources
+- Propagating logs from the WF isolate has similar implications
+
+### Alternatives
```

Comment:
Other options we discussed on the call:

I wrote the paragraphs below this one first, but after writing them, I have the feeling that the "right" thing to do here is actually always have lang be the initiator of a span (which makes sense) and then core simply attaches it's additional spans/events/metrics/etc into the "context" provided by lang, and that context can be returned by each API call. Unfortunately this makes our APIs super ugly purely for the sake of opentelemetry which I don't like at all. So, I'd still prefer a solution that allows us to include the span/context in our protos and then propagate results out-of-band.

One option would be to have core and lang share a socket (in memory, or not, whatever) which actually uses the gRPC protocol in streaming mode to have core constantly pump tracing data to the lang side. Probably using https://docs.rs/opentelemetry/0.14.0/opentelemetry/propagation/index.html or alternatively https://docs.rs/opentelemetry/0.14.0/opentelemetry/sdk/export/trace/trait.SpanExporter.html -- it's not entirely clear from the docs what makes more sense. Will need to experiment. The span exporter seems slightly more obvious, buuuut, the propagators could potentially better suit our needs if we can somehow just pass them along with every API call, core can use them to inject everything it's doing and then lang will end up with everything on it's side.

Another would be to simply define the lang side as an opentelemetry "processor" (I think? Their concepts are not super well explained) https://opentelemetry.io/docs/concepts/data-collection/ -- but essentially just make lang a receiver of exported data from core, where it would then combine it with it's own data and export everything to whatever backend(s) the user has registered. Problem is, after a little research, exporting the data this way using OTLP for example looks pretty easy https://docs.rs/opentelemetry-otlp/0.7.0/opentelemetry_otlp/ but receiving it looks much harder as there doesn't appear to be much in the way of libraries for that part which isn't terribly surprising. We could implement it ourselves, but at that point the other options are probably easier.

Reply:
Another thing we had considered is that core respects the sampling flags in the trace context https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampling 
This should probably come out of the box if we use the library in a standard way.

Reply:
See also this comment about missing unique IDs for activations https://github.com/temporalio/sdk-core/pull/125/files#r638239225

Reply:
I'm strongly leaning towards writing a custom propagator or exporter that will push spans from Core to Lang.

Reply:
Yeah I think that's certainly the first avenue to explore. And yeah I suppose there are some circumstances where it would make sense for core to initiate a span, but they could always be child spans of something initiated by lang. Heartbeats for example initiated by HB call in lang and then have child spans for every individual actual heartbeat request, but that possibly ends up being unnecessary in practice.